Archive/Examples/SumOracle.lean

```diff
@@ -0,0 +1,63 @@
+import Mathlib.Computability.QueryComplexity.Defs
+import Mathlib.Data.Fintype.BigOperators
+import Mathlib.Data.ENat.Lattice
+import Mathlib.Algebra.FreeMonoid.Basic
+
+namespace ArithOracle
+
+inductive AddQuery (α : Type*)
+| add (x y : α)
+| zero
+
+open QueryComplexity.Comp (query)
+
+abbrev Comp (α : Type*) := QueryComplexity.Comp (AddQuery α) (fun _ => α)
+
+variable {α β}
+
+/-- Sum the elements of `A`. -/
+def sum {n : ℕ} (A : Fin n → α) : Comp α α := do
+  List.ofFn A |>.foldrM
+    (fun x a => query (.add x a))
+    (← query .zero)
+
+/-- The cost and output of a computation in `Comp`. -/
+def Comp.run [Zero α] [Add α] (f : Comp α β) : β × ℕ :=
+  StateT.run (s := 0) (m := Id) <|
+    f.runM fun
+      | .add x y => do
+        modify fun n => n + 1
+        return x + y
+      | .zero => return 0
+
+/-- `sum` computes the sum of a list, using `n` additions. -/
+theorem sum_run [Zero α] [Add α] (n : ℕ) (A : Fin n → α) :
+    (sum A).run = ((List.ofFn A).sum, n) := by
+  induction n with
+  | zero =>
+    simp [sum, Comp.run]
+  | succ n hn =>
+    simp_all [sum, Comp.run]
+
+/-- The computation must use at least `n` additions to sum `n` elements. -/
+-- TODO: is this false because `sum` can be `if α = ℕ then cheat else dont_cheat`?
```

Comment:
@girving, am I right in thinking that this just isn't possible to state / prove with the `Comp` machinery?

Reply:
Correct: since the data is passed in outside the oracle, you don’t even need the if: you should be able to always cheat.

Reply:
The data isn't fully passed in outside the oracle; the `Add` instance is not visible to the computation,  only the carrier type is.